DR-1842 Create storage accounts when creating Azure backed datasets

nmalfroy opened the pull request:
First off, apologies for the sheer size of this PR.  I did break it into smaller commits to make it easier to review but this seemed like a quicker way to make progress on this. (also worth noting, 4K of the lines are from a single test file...I may try to find a smaller one...)

This PR:
- Creates the tables to store deployed application, storage account metadata and dataset / storage account relationships as well as columns to link to snapshots and datasets
- All of the dao methods and support CRUD on theses values
- The service layer to combine CRUD on metadata as well as creation deletion of storage resources
- A whole lot of unit testing
- Added a bunch of flight steps on dataset create / delete and profile delete to interact with the storage accounts
  -  Worth noting that deleting storage accounts are done in one transactional step.  I think it makes it easier to maintain
- An integration test at the end to test the whole thing (and to build off of as we add file ingest)

Also worth noting: I made the object mapper that we use accessible via a static method (since in some cases, I couldn't autowire it)

nmalfroy committed:
Remove unnecessary cast

nmalfroy committed:
Add new database definitions

nmalfroy committed:
Fix issue where results from newly created profiles had null values for Azure UUIDs

nmalfroy committed:
Create java POJOs for application deployments and storage accounts

nmalfroy committed:
Add dao layer for application deployments and storage accounts

nmalfroy committed:
Make default object mapper also available as a static variable

nmalfroy committed:
Create ApplicationDeployment service class

nmalfroy committed:
Create StorageAccount service class

nmalfroy committed:
Add new application deployment link to dataset POJO

nmalfroy committed:
Add suport for Azure regions in dataset creation

nmalfroy committed:
Add dao for linking storage accounts to datasets

nmalfroy committed:
Update resource service as entry point for creating storage accounts from flights

nmalfroy committed:
Add support methods to handle deleting a storage account

nmalfroy committed:
Add profile delete flight Azure steps

nmalfroy committed:
Add dataset create flight Azure steps

nmalfroy committed:
Add dataset delete flight Azure steps

nmalfroy committed:
Add integration test config and fixture method to create azure billing profile

nmalfroy committed:
Add an option for integration tests to be able to get users deterministically

we need this since we depend on the user being set in the application deployment

nmalfroy committed:
Add an integration test to test creation of Azure datasets

nmalfroy committed:
Appease the checkstyle gods

nmalfroy committed:
Add cloud type explicitly to unit/connected test dataset creation requests

nmalfroy committed:
Default to GCP if no cloud platform specified in dataset creation

For backwards compatibility

pshapiro4broad commented on src/main/java/bio/terra/app/configuration/ApplicationConfiguration.java:
```diff
@@ -22,6 +22,14 @@
 @ConfigurationProperties(prefix = "datarepo")
 public class ApplicationConfiguration {
+    /**
+     * Adding a static variable that can be accessed from the testing code since the MockitoJUnitRunner doesn't seem
+     * to allow you to autowire and having the same object mapper used in the tests is important
```
Do you need the same object, or would it be OK to use another object that was created using the same registered modules?

If you have a test that doesn't use injection, you can create same object mapper by using `new ApplicationConfiguration().objectMapper()`.

nmalfroy replied:
Ah fair point...I just wanted to make sure that there was a common way to create an ObjectMapper with the same config.  It feels roundabout to get an object mapper that way, but happy to change it

pshapiro4broad commented on src/main/java/bio/terra/app/model/AzureStorageAccountSkuType.java:
```diff
@@ -0,0 +1,37 @@
+     * or null if not found
+     */
+    public static AzureStorageAccountSkuType fromAzureName(String name) {
+        SkuName skuName = SkuName.fromString(name);
```
This could go either way, but maybe the caller should be handling the conversion to `SkuName`.

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/DatasetJsonConversion.java:
```diff
@@ -82,10 +87,23 @@ public static Dataset datasetRequestToDataset(DatasetRequestModel datasetRequest
 
     public static GoogleRegion getRegionFromDatasetRequestModel(DatasetRequestModel datasetRequestModel) {
-        return GoogleRegion.fromValueWithDefault(datasetRequestModel.getRegion());
+        // Default to GCP if no platform was specified
+        if (datasetRequestModel.getCloudPlatform() == null ||
```
Can this be `null` in practice? The openapi spec says that it has a default, so it should always be set before this point.

If you do need to handle `null`, you should use `DEFAULT_CLOUD_PLATFORM` here, not hardcode `GCP`.

nmalfroy replied:
Yeah, sadly, it can be...fair point about using default cloud platform

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/DatasetJsonConversion.java:
```diff
@@ -82,10 +87,23 @@ public static Dataset datasetRequestToDataset(DatasetRequestModel datasetRequest
     }
 
     public static GoogleRegion getRegionFromDatasetRequestModel(DatasetRequestModel datasetRequestModel) {
```
I think what I'd like to see is this method moved into a class that wraps `CloudPlatform`. If `CloudPlatform` wasn't auto generated, we could use an enum method to implement the different features. Using inheritance to implement the wrapper would also limit the need to switch on `CloudPlatform` to the factory method that creates the wrapper.

Although maybe for now, we could start with moving all the platform switch code into a separate object.

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/DatasetJsonConversion.java:
```diff
@@ -100,6 +118,18 @@ public static GoogleRegion getRegionFromDatasetRequestModel(DatasetRequestModel
+            Arrays.stream(AzureCloudResource.values()).map(resource -> new AzureStorageResource(null,
+                resource,
+                region)),
+            createGcpStorageResourceValues(datasetRequestModel).stream()).flatMap(s -> s)
```
Instead of `flatMap()` you can directly concatenate the two streams.
```suggestion
        return Stream.concat(
            Arrays.stream(AzureCloudResource.values()).map(resource -> new AzureStorageResource(null,
                resource,
                region)),
            createGcpStorageResourceValues(datasetRequestModel).stream())
```

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/DatasetJsonConversion.java:
```diff
@@ -100,6 +118,18 @@ public static GoogleRegion getRegionFromDatasetRequestModel(DatasetRequestModel
+    private static List<? extends StorageResource<?, ?>> createAzureStorageResourceValues(
+        DatasetRequestModel datasetRequestModel) {
+        final AzureRegion region = getAzureRegionFromDatasetRequestModel(datasetRequestModel);
+        // TODO: once we no longer require GCP resources to back Azure datasets, stop concatenating
```
It would be better to put this information in a JIRA issue than as a TODO in the code, as we currently have no process to review and monitor TODOs in our code.

nmalfroy replied:
I actually kind of like having it here.  This is going to be a gradual pull back as we move more and more of the regional resources to azure for azure backed datasets

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/DatasetRequestValidator.java:
```diff
@@ -377,18 +381,32 @@ private void validateRegion(DatasetRequestModel datasetRequest, Errors errors) {
+                                "Valid regions for " +
+                                    CloudPlatform.GCP +
+                                    " are: " + String.join(", ", SUPPORTED_GOOGLE_REGIONS));
+                        }
```
This code fragment is duplicated three times, refactoring can eliminate the duplication. Something like
```
    private void ensureValidRegion(CloudPlatform platform, String region, List<String> supportedRegions, Errors errors) {
        if (!supportedRegions.contains(region.toLowerCase())) {
            errors.rejectValue("region", "InvalidRegionForPlatform",
                "Valid regions for " + platform +  " are: " + String.join(", ", supportedRegions));
        }
    }
```

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/DatasetRequestValidator.java:
```diff
@@ -43,6 +44,9 @@
         Arrays.stream(GoogleRegion.values()).map(GoogleRegion::getValue).collect(Collectors.toUnmodifiableList());
 
+    private static final List<String> SUPPORTED_AZURE_REGIONS =
+        Arrays.stream(AzureRegion.values()).map(AzureRegion::toString).collect(Collectors.toUnmodifiableList());
```
Why is this `toString()` and `GoogleRegion` is `toValue()`?

Also, these should probably be static members of the enums `AzureRegion` and `GoogleRegion`. If you move them, I would remove `GOOGLE` and `AZURE` from their names as it would be redundant then.

pshapiro4broad replied:
Pretty sure this code can be deleted now that it was moved to `AzureRegion.SUPPORTED_REGIONS`

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/DatasetStorageAccountDao.java:
```diff
@@ -0,0 +1,174 @@
+
+import static bio.terra.common.DaoUtils.retryQuery;
+
+/**
```
This isn't a javadoc comment so it shouldn't start with `/**`
```suggestion
/*
```

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/DatasetStorageAccountDao.java:
```diff
@@ -0,0 +1,174 @@
+ *
+ * An ingest either creates the row with a value of 1 or it increments the value. The undo step always
+ * decrements the value. That allows parallel ingests. On failure, we do not worry about trying to
+ * removed the row.
```
typo
```suggestion
 * remove the row.
```

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/DatasetStorageAccountDao.java:
```diff
@@ -0,0 +1,174 @@
+    }
+
+    @Transactional(propagation = Propagation.REQUIRED, isolation = Isolation.SERIALIZABLE)
+    public void decrementDatasetStorageAccountLink(UUID datasetId, UUID storageAccountResourceId) {
```
This method appears to be currently unused, and is not tested.

nmalfroy replied:
Yeah, that's copied over from the bucket equivalent.  I'll add some testing for these two methods

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/DatasetStorageAccountDao.java:
```diff
@@ -0,0 +1,174 @@
+    }
+
+    @Transactional(propagation = Propagation.REQUIRED, isolation = Isolation.SERIALIZABLE)
+    int datasetStorageAccountSuccessfulIngestCount(UUID datasetId, UUID storageAccountResourceId) {
```
This is also currently unused and untested.

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/DatasetStorageAccountDao.java:
```diff
@@ -0,0 +1,174 @@
+    }
+
+    private static class UuidMapper implements RowMapper<UUID> {
+        private String columnLabel;
```
can be `final`
```suggestion
        private final String columnLabel;
```

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/flight/create/CreateDatasetCreateStorageAccountLinkStep.java:
```diff
@@ -0,0 +1,53 @@
+
+            datasetStorageAccountDao.createDatasetStorageAccountLink(datasetId, storageAccountId, false);
+        } else {
+            logger.info("Not an Azure backed dataset so no action to take");
```
Would it be better to not add the step in the first place, when creating the flight?

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/flight/create/CreateDatasetCreateStorageAccountLinkStep.java:
```diff
@@ -0,0 +1,53 @@
+
+    @Override
+    public StepResult undoStep(FlightContext context) throws InterruptedException {
+        logger.debug("Dataset / Storage account link creation failed. Deleting link metadata.");
```
Seems like this should check `datasetRequestModel.getCloudPlatform() == CloudPlatform.AZURE` before doing the undo operation?

pshapiro4broad replied:
This step is now only added if `isAzure()` is true.

nmalfroy committed:
Object mapper is no longer a static variable

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/flight/create/CreateDatasetGetOrCreateStorageAccountStep.java:
```diff
@@ -0,0 +1,61 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CreateDatasetGetOrCreateStorageAccountStep implements Step {
```
Does this class name mean that it can only be used inside a CreateDataset flight? If we had another flight that needs to get or create a storage account, would its step be different from this one?

pshapiro4broad replied:
I still have questions about how we name steps.. maybe most steps are very flight specific? I guess given how inputs and outputs are handled (fixed keys in a hash map) they would have to be.

nmalfroy committed:
Convert usage of String for IDs to UUIDs

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/flight/delete/DeleteDatasetDeleteStorageAccountsStep.java:
```diff
@@ -0,0 +1,50 @@
+        boolean isAzureDataset = dataset.getStorage().stream()
+            .anyMatch(s -> s.getCloudPlatform() == CloudPlatform.AZURE);
+        if (isAzureDataset) {
+            logger.info("Deleting a storage account for Azure backed dataset");
```
Is it typical to `info` log in our steps?

nmalfroy replied:
Yeah, we do some of that.  It certainly happens within the dao methods.  Makes figuring out what's going on with flights easier

pshapiro4broad commented on src/main/java/bio/terra/service/profile/ProfileDao.java:
```diff
@@ -98,8 +98,12 @@ public BillingProfileModel createBillingProfile(BillingProfileRequestModel profi
             .cloudPlatform(CloudPlatform.valueOf(keyHolder.getString("cloud_platform")))
-            .tenantId(keyHolder.getString("tenant_id"))
-            .subscriptionId(keyHolder.getString("subscription_id"))
+            .tenantId(Optional.ofNullable(keyHolder.getField("tenant_id", UUID.class))
```
This is ok, but it would be better to have `getField()` return `Optional<T>` instead of `T`. I'm now thinking every time we write `Optional.ofNullable()` and it's not in a `return` statement, we should try to update the code to move it to a `return`, if it's in our source code.

pshapiro4broad commented on src/main/java/bio/terra/service/resourcemanagement/AzureDataLocationSelector.java:
```diff
@@ -0,0 +1,52 @@
+
+@Primary
+@Component
+public class AzureDataLocationSelector {
```
Why is this an injectable? Both methods can be static.

nmalfroy replied:
To make it match with the GCP based locator

pshapiro4broad commented on src/main/java/bio/terra/service/resourcemanagement/AzureDataLocationSelector.java:
```diff
@@ -0,0 +1,52 @@
+     * @param size The number of characters to generate
+     * @return An n-character unique string based on what was passed into seed
+     */
+    private String armUniqueString(String seed, int size) {
```
This code appears to be duplicated in `AzureResourceConfigurationTest`

nmalfroy replied:
Yeah, I'll pull out the one from the testing

pshapiro4broad commented on src/main/java/bio/terra/service/resourcemanagement/ResourceService.java:
```diff
@@ -103,6 +118,86 @@ public GoogleBucketResource getOrCreateBucketForFile(Dataset dataset,
+     * @return a reference to the bucket as a POJO AzureStorageAccountResource
+     * @throws CorruptMetadataException in two cases.
+     * <ol>
+     *     <le>if the storage account already exists, but the metadata does not AND
+     *     the application property allowReuseExistingBuckets=false.</le>
+     *     <le>if the metadata exists, but the storage account does not</le>
```
Not sure what `le` is but I don't think it will work. I'd use `li`.
```suggestion
     *     <li>if the storage account already exists, but the metadata does not AND
     *     the application property allowReuseExistingBuckets=false.</li>
     *     <li>if the metadata exists, but the storage account does not</li>
```
Also I'd use `ul` not `ol` but that's up to you.

nmalfroy replied:
ah, copy pasta.  fixed elsewhere too

pshapiro4broad reviewed (commented):
Comments on the first 20 files.. Am just getting to the big changes, I think.

nmalfroy committed:
DaoKeyHolder.getField now returns an optional

nmalfroy committed:
Add method to do the switching on cloud provider

nmalfroy committed:
Use a single armUniqueString method

nmalfroy committed:
Fix javadoc

tlangs commented on src/main/java/bio/terra/common/CloudUtil.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    /**
+     * Execute the code in the appropriate supplier based on the cloud platform used.
```
```suggestion
     * Execute the code in the appropriate Runnable based on the cloud platform used.
```

nmalfroy replied:
I ended up getting rid of this class

tlangs commented on src/main/java/bio/terra/common/CloudUtil.java:
```diff
@@ -0,0 +1,64 @@
+     */
+    public static void cloudExecute(CloudPlatform cloudPlatform,
+                                    Runnable gcpRunnable,
+                                    Runnable azureRunnable) {
```
Not necessarily an issue. I associate `Runnable`s with threads, and from an method API perspective, I might assume that `cloudExecute` might run things within separate threads if I didn't look at the method definition. Unfortunately, Java doesn't provide a Functional Interface for a no-arg, no return function, so we'd have to make our own `Action`-type interface. IDK if its worth the effort, but something to think about.

tlangs commented on src/main/java/bio/terra/common/DaoKeyHolder.java:
```diff
@@ -5,16 +5,17 @@
 
     public UUID getId() {
-        return getField("id", UUID.class);
+        return getField("id", UUID.class).orElse(null);
```
Should this be `orElseThrow()`? I can't think of an instance where we'd want to DB to not generate an ID.

nmalfroy replied:
I wasn't quite sure.  Wanted to try leaving for now to leave the current behavior but I may change this in a followon

tlangs commented on src/main/java/bio/terra/service/profile/flight/delete/ProfileDeleteFlight.java:
```diff
@@ -39,9 +49,24 @@ public ProfileDeleteFlight(FlightMap inputParameters, Object applicationContext)
         addStep(new DeleteProfileMarkUnusedProjects(resourceService, profileId));
+        CloudUtil.cloudExecute(
+            cloudPlatform,
+            () -> { },
```
I'm seeing this pattern repeated a lot. Making a `CloudUtil.NOTHING` constant would clean things up a little and reduce repetition of symbols.

nmalfroy replied:
Good point.  I was also thinking we may also want an "Only run this on Azure" or "Only run this on GCP" type method

nmalfroy replied:
I ended up changing the way this works based on Phil's feedback

tlangs commented on src/main/java/bio/terra/service/resourcemanagement/ResourceService.java:
```diff
@@ -103,6 +118,86 @@ public GoogleBucketResource getOrCreateBucketForFile(Dataset dataset,
+            storageAccountName = storageAccount.getName();
+
+            if (storageAccountsForBillingProfile.size() > 1) {
+                logger.warn("Found more than one storage account associated with this dataset and billing profile");
```
If there should never be more than one billing account, should we throw instead of just warning?

nmalfroy replied:
I actually think there might be legit usecases for this (e.g. if you're ingesting with different billing profiles...I'm not sure if that should be a failure...)

tlangs replied:
🤷

tlangs commented on src/main/java/bio/terra/service/resourcemanagement/azure/AzureResourceDao.java:
```diff
@@ -0,0 +1,424 @@
+            new MapSqlParameterSource().addValue("azure_application_deployment_name", applicationName);
+        return retrieveApplicationDeploymentBy(sqlApplicationRetrieveByDepName, params);
+    }
+    @Transactional(propagation = Propagation.REQUIRED, readOnly = true)
```
Do we need `isolation` options in these `@Transactional` annotations? 😉

nmalfroy replied:
haha we do!

pshapiro4broad commented on src/main/java/bio/terra/common/CloudUtil.java:
```diff
@@ -0,0 +1,64 @@
+/**
+ * Utility methods for working across cloud providers
+ */
+public class CloudUtil {
```
To me, this doesn't solve the main problem, which is having to put different code for different platforms throughout our code base.

Instead of making it easier to `switch`, we should have an interface for different cloud operations and use subclasses to implement the operations for different platforms.

nmalfroy replied:
Oh, I think I see.  I have to say in that case I don't agree.  We're switching based on values coming in...I don't see how we avoid having to to figure out the action.  We could add it to a subclass but I don't think that buys you much short have needing to go through more classes to understand the code.  You need to end up needing to instantiate these subclasses using a switch anyways...so I'm not totally sure what it buys you

nmalfroy replied:
Ah, I see from your other PR...ok.  That makes more sense.  My concern there is in the size that this class might get.  Seems like having a single class do the branching is going to get massive pretty quickly.  It's also worth noting that we already have some abstraction at the individual class types (e.g. CloudStorage, Region, etc.).  That being said, I do like the idea of having some common methods like that in a common interface and having some way to instantiate it based on common objects...

nmalfroy replied:
Cool, so I ended up incorporating your suggested implementation (couple super small changes around branching in the flights) but the last commit in the PR should have the thing

tlangs commented on src/main/java/bio/terra/service/resourcemanagement/azure/AzureResourceDao.java:
```diff
@@ -0,0 +1,424 @@
+            "    AND storage_account_resource.id = dataset_storage_account.storage_account_resource_id" +
+            "    AND dataset_storage_account.successful_ingests > 0) AS sacnt" +
+            " FROM application_deployment_resource AS app " +
+            " WHERE app" +
+            ".profile_id = :profile_id) AS X";
```
Weird line break.
```suggestion
            " WHERE app.profile_id = :profile_id) AS X";
```

nmalfroy replied:
yep

tlangs commented on src/main/java/bio/terra/service/resourcemanagement/azure/AzureResourceDao.java:
```diff
@@ -0,0 +1,424 @@
+                .refCount(rs.getLong("refcnt")));
+
+        // If the profile is in use by any project, we bail here.
+        long totalRefs = 0;
+        for (AppRefs ref : appRefs) {
+            logger.info("Profile app reference appId: {} refCount: {}", ref.getAppId(), ref.getRefCount());
+            totalRefs += ref.getRefCount();
+        }
```
Take it or leave it, but Streams are pretty cool!

```suggestion
        long totalRefs = appRefs.stream()
            .peek(r -> logger.info("Profile app reference appId: {} refCount: {}", r.getAppId(), r.getRefCount()))
            .mapToLong(AppRefs::getRefCount).sum();
        }
```

nmalfroy replied:
Yeah, I agree but I want to be careful not to go to overboard with lambdas...it can hurt readability a little bit

tlangs replied:
I actually find lambdas to be more readable in a lot of cases. Especially with streams, the operations are more sequential and don't modify a variable that I have to keep track of in my head. But these are stylistic choices and not necessary for merging

nmalfroy replied:
Yeah, I don't disagree with you. Maybe we can start leaning on them more as we move forward

tlangs commented on src/main/java/bio/terra/service/resourcemanagement/azure/AzureResourceDao.java:
```diff
@@ -0,0 +1,424 @@
+
+    private AzureStorageAccountResource retrieveStorageAccountBy(String sql, MapSqlParameterSource params) {
+        List<AzureStorageAccountResource> storageAccountResources =
+            jdbcTemplate.query(sql, params, (rs, rowNum) -> {
```
If we only want one, why not use `queryForObject`?

nmalfroy replied:
I think it's possible because there are cases where it could be more than one?

tlangs replied:
But later down, we throw if there's more than one.

nmalfroy replied:
Ah wait, I see why we don't use queryForObject. We don't want to throw if nothing is found and queryForObject does that

tlangs replied:
ah gotcha.

tlangs commented on src/main/java/bio/terra/service/resourcemanagement/azure/AzureStorageAccountService.java:
```diff
@@ -0,0 +1,325 @@
+            BillingProfileModel billingProfile =
+                profileDao.getBillingProfileById(storageAccountResource.getProfileId());
+            // throw an exception if the storage account does not already exist
+            StorageAccount storageAccount = getCloudStorageAccount(billingProfile, storageAccountResource);
```
Really appreciating Phil's comments about returning `Optional`s instead of `null` from a lot of our functions.

snf2ye reviewed (approved):
Thanks for walking us through these changes!

nmalfroy committed:
Transaction annotation fix

nmalfroy committed:
Switch to use CloudPlatformWrapper

Also fix failure when deleting a dataset that was already deleted.  We want the job to launch but then not do anything

nmalfroy committed:
Make profile not 404 on the request if it doesn't exist

- The fligh handles it

pshapiro4broad commented on src/main/java/bio/terra/app/model/AzureRegion.java:
```diff
@@ -77,6 +80,9 @@
     public static final AzureRegion DEFAULT_AZURE_REGION = AzureRegion.CENTRAL_US;
 
+    public static final List<String> SUPPORTED_REGIONS =
+        Arrays.stream(AzureRegion.values()).map(AzureRegion::toString).collect(Collectors.toUnmodifiableList());
```
minor improvement (?)
```suggestion
        Arrays.stream(values()).map(AzureRegion::toString).collect(Collectors.toUnmodifiableList());
```

samanehsan commented on src/main/java/bio/terra/service/dataset/DatasetJsonConversion.java:
```diff
@@ -61,15 +56,10 @@ public static Dataset datasetRequestToDataset(DatasetRequestModel datasetRequest
-            throw new UnsupportedOperationException(cloudPlatform + " is not a recognized Cloud Platform");
-        }
+        final List<? extends StorageResource<?, ?>> storageResources =
+            cloudPlatform.createStorageResourceValues(datasetRequest);
```
I like the abstraction here!

pshapiro4broad commented on src/main/java/bio/terra/common/DaoKeyHolder.java:
```diff
@@ -5,16 +5,17 @@
 
 public class DaoKeyHolder extends GeneratedKeyHolder {
 
     public UUID getId() {
```
I was thinking these methods would all be changed to return `Optional` e.g. `Optional<UUID>`. Although that would push the null check into the caller. Maybe something for a future PR.

samanehsan commented on src/main/java/bio/terra/service/dataset/DatasetRequestValidator.java:
```diff
@@ -40,8 +40,8 @@
 
-    private static final List<String> SUPPORTED_GOOGLE_REGIONS =
-        Arrays.stream(GoogleRegion.values()).map(GoogleRegion::getValue).collect(Collectors.toUnmodifiableList());
+    private static final List<String> SUPPORTED_AZURE_REGIONS =
```
We can remove this, right?

nmalfroy replied:
Ah yes!

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/DatasetService.java:
```diff
@@ -133,9 +142,20 @@ public EnumerateDatasetModel enumerate(int offset,
 
     public String delete(String id, AuthenticatedUserRequest userReq) {
         String description = "Delete dataset " + id;
+        CloudPlatform platform;
```
can be `final`
```suggestion
        final CloudPlatform platform;
```

nmalfroy replied:
That doesn't seem to work with try/catch instantiation :(

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/flight/create/DatasetCreateFlight.java:
```diff
@@ -33,26 +36,45 @@ public DatasetCreateFlight(FlightMap inputParameters, Object applicationContext)
         IamProviderInterface iamClient = (IamProviderInterface) appContext.getBean("iamProvider");
         ConfigurationService configService = (ConfigurationService) appContext.getBean("configurationService");
         ProfileService profileService = (ProfileService) appContext.getBean("profileService");
+        AzureDataLocationSelector azureDataLocationSelector =
+            (AzureDataLocationSelector) appContext.getBean("azureDataLocationSelector");
+        DatasetStorageAccountDao datasetStorageAccountDao =
+            (DatasetStorageAccountDao) appContext.getBean("datasetStorageAccountDao");
```
This is a pattern in our code we can improve. I believe these beans don't need to be looked up by name; if so then the code could use `getBean(Class.class)`. If we do need to look up by name, we can at least use the API that also accepts a class argument.
```suggestion
        AzureDataLocationSelector azureDataLocationSelector =
            appContext.getBean("azureDataLocationSelector", AzureDataLocationSelector.class);
        DatasetStorageAccountDao datasetStorageAccountDao =
            appContext.getBean("datasetStorageAccountDao", DatasetStorageAccountDao.class);
```

nmalfroy replied:
Yeah, I've been wanting to do a wholesale change over to the class instantiation

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/flight/delete/DeleteDatasetDeleteStorageAccountsStep.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.UUID;
+
+public class DeleteDatasetDeleteStorageAccountsStep implements Step {
+    private static Logger logger = LoggerFactory.getLogger(DeleteDatasetDeleteStorageAccountsStep.class);
```
can be `final`
```suggestion
    private static final Logger logger = LoggerFactory.getLogger(DeleteDatasetDeleteStorageAccountsStep.class);
```

pshapiro4broad commented on src/main/java/bio/terra/service/job/JobMapKeys.java:
```diff
@@ -10,6 +10,7 @@
     STATUS_CODE("status_code"),
     AUTH_USER_INFO("auth_user_info"),
     SUBJECT_ID("subjectId"),
+    CLOUD_PLATFORM("cloudPlatform"),
```
This should really not be an enum (compare with `FileMapKeys`, `LoadMapKeys`, etc). more tech debt I suppose.

pshapiro4broad commented on src/main/java/bio/terra/service/profile/ProfileService.java:
```diff
@@ -115,11 +118,21 @@ public String updateProfile(BillingProfileUpdateModel billingProfileRequest,
         return jobService
             .newJob(description, ProfileDeleteFlight.class, null, user)
             .addParameter(ProfileMapKeys.PROFILE_ID, id)
+            .addParameter(JobMapKeys.CLOUD_PLATFORM.getKeyName(), platform)
```
I wonder if it would work better to use `CloudPlatformWrapper` here instead of `CloudPlatform`.

nmalfroy replied:
I was wondering about this.  These get serialized into the DB so I like the idea of keeping this smaller

samanehsan reviewed (approved):
This looks great! Lgtm 🌤️
